<launch>
    <arg name="xacro_path"   default="$(find ur5_allegro_moveit)/models/ur5_allegro.xacro" />
    <arg name="urdf_args"   default=""/>
    <arg name="load_xacro"   default="true"/>
    <arg name="viz"   default="true"/>
    <arg name="start_leap"   default="true"/>
    <arg name="leap_topic"   default="/leap_motion/leap_filtered"/>
    <arg name="left_handed"   default="False"/>
    <arg name="joint_state_gui"   default="False"/>
    <arg name="debug_mode"   default="False"/>
    <arg name="sim"   default="True"/>
    <arg name="render"   default="True"/>
    <arg name="world_name"   default="calibration-world.xml"/>
    
    <group if="$(arg sim)"> 
        <include file="$(find allegro_mujoco)/launch/mujoco_ros_communication.launch">
                    <arg name="rviz"   value="false"/>
                    <arg name="render" value="$(arg render)"/>
                    <arg name="load_robot" value="ur5_allegro"/>
                    <arg name="world_model_file" type="str" value="$(find allegro_mujoco)/config/worlds/$(arg world_name)" />
        </include>
    </group>
    

    <group>
        <include file="$(find allegro_hand_kdl)/launch/gravity_compensate.launch">
          <!-- <arg name='TORQUE_TOPIC' value='/allegroHand_0/torque_cmd'/> -->
        </include>
        
        <include file="$(find allegro_hand_kdl)/launch/pose_action_server.launch">
            
            <arg name="NUM" value="0"/>
            <!-- <arg name="HAND" value="right"/> -->
            <arg name="HAND" value="right" if="$(eval arg('left_handed') == false)"/>
            <arg name="HAND" value="left" if="$(eval arg('left_handed') == true)"/>
            <arg name="TIME_LIMIT" value="4.0" />
            <arg name="TORQUE_TOPIC" value="position_torque"/>
            
        </include>
        
        
        <include file="$(find allegro_hand_kdl)/launch/combine_torques.launch"/>
    </group>
    

    <!-- From HERE on we have only teleop and leap nodes -->

    <group if="$(arg start_leap)">
        <include file="$(find teleoperation_ur5_allegro_leap)/launch/leap_start.launch">
            <arg name="leap_camera_mode"   value="False"/>
            <arg name="leap_hand_root"   value="wrist"/>
            <arg name="leap_tf_hand_parent" value="leap_hands" />
            <arg name="leap_finger_parent" value="hand_root" />

            <!-- <arg name="leap_hand_tf" value="0 0 0 0.7071067811865475 0.0 0.0 0.7071067811865476 world leap_hands"/> -->
            <!-- <arg name="leap_hand_tf" default="0 0 0 -0.707 0.707 0.000 0.000 hand_root leap_hands"/> -->
            <arg name="leap_hand_tf" default="0 0 0 0.5000000532124816 0.5000000532124816 0.5 0.5000000532124816 world leap_hands"/>
            <!-- <arg name="leap_right_fingers_tf" value="0 0 0 0.5 -0.5 0.5 0.5 hand_root right_leap_fingers"/> -->
            <arg name="leap_right_fingers_tf" value="0 0 0 0 0 0 1 hand_root right_leap_fingers"/>
            <arg name="leap_left_fingers_tf" value="0 0 0 0.5 0.5 0.5 0.5 world left_leap_fingers"/>
        </include>
    </group>
    

    <node pkg="teleoperation_ur5_allegro_leap" type="leap2allegro.py" name="allegro_teleop" output="screen">
        <param name="left_hand" type="bool" value="false"/>
        <param name="consumption_rate" type="double" value="60" />
        <param name="leap_topic" type="str" value="$(arg leap_topic)" />
    </node>

    <!-- <node pkg="teleoperation_ur5_allegro_leap" type="teleop_gui_node.py" name="teleop_gui" output="screen"> -->
    <node pkg="teleoperation_ur5_allegro_leap" type="full_teleop_gui.py" name="teleop_gui" output="screen">
        <param name="arm_teleop" type="bool" value="false"/>
        <param name="hand_teleop" type="bool" value="true" />
    </node>

    <group if="$(arg viz)">
        <node name="teleop_rviz" pkg="rviz" type="rviz" args="-d $(find teleoperation_ur5_allegro_leap)/rviz/teleop_viz.rviz" />
    </group>

    <include file="$(find haptic_glove_ros)/launch/opto2glove.launch">
            
    </include>
</launch>